spmd/tensor/parallel/api.py

```diff
@@ -14,6 +14,176 @@
+
+    assert device_mesh is not None, "_replicate_input_1d: device mesh is not 1D"
+
+    replicate = [Replicate()]
+    if isinstance(torch.Tensor, tensor):
+        return DTensor.from_local(tensor, device_mesh, replicate)
+    elif isinstance(DTensor, tensor):
+        return tensor.redistribute(device_mesh, replicate)
+    else:
+        raise RuntimeError(
+            "Input to _replicate_input_1d must be either torch.Tensor or DTensor!"
+        )
+
+
+def MakeInputReplicated() -> functools.partial[DTensor]:
+    # This function generates input handler for 1-D mesh device only
+    return functools.partial(_replicate_input_1d)
+
+
+def _replicate_output_1d(tensor: DTensor) -> torch.Tensor:
+    assert (
+        isinstance(tensor, DTensor) and tensor.device_mesh.ndim == 1
+    ), "_replicate_output_1d: device mesh is not 1D"
+    return replicate_output(tensor)
+
+
+def MakeOutputReplicated() -> functools.partial[torch.Tensor]:
+    # This function generates input handler for 1-D mesh device only
+    return functools.partial(_replicate_output_1d)
+
+
+@dataclass
+class ParallelStyle(object):
+    """
+    The parallel style user wants the module or submodule to be sharded.
+    Users can extend this class to build their own parallel style with customized input/output preparations.
+    """
+
+    _prepare_input: Callable[
+        [Union[torch.Tensor, DTensor], Optional[DeviceMesh]], DTensor
+    ]
+    _prepare_output: Callable[[DTensor], Union[Tensor, DTensor]]
+
+
+class RowwiseParallel(ParallelStyle):
+    """
+    Partitioning the row of a module. We assume the input to be a Shard(-1) DTensor and output to be a replicated DTensor.
+    """
+
+    def __init__(self):
+        super().__init__(MakeInputShard(-1), MakeOutputReplicated())
+
+
+class ColwiseParallel(ParallelStyle):
+    """
+    Partitioning the column of a tensor or module. We assume the input to be a Replicated DTensor and output to be a Shard(-1) DTensor.
+    """
+
+    def __init__(self):
+        super().__init__(MakeInputReplicated(), None)
+
+
+class PairwiseParallel(ParallelStyle):
+    """
+    We concatenate colwise and rowwise styles as a fixed pair like what Megatron-LM(https://arxiv.org/abs/1909.08053) is doing. We assume both input and output to a Replicated DTensor. We now only support Multihead Attention, MLP and transformer for this style.
+    We also need to assume the input is a nn.Multihead Attention, nn.Transformer or even-number layers of nn.Linear for now.
+    """
+
+    def __init__(self):
+        super().__init__(MakeInputReplicated(), MakeOutputReplicated())
+
+
+def _parallelize_linear_1d(
+    module: nn.Module,
+    parallel_style: ParallelStyle = ColwiseParallel(),
+    device_mesh: DeviceMesh = None,
+    tp_mesh_dim: int = 0,  # TODO: remove this param???
+) -> None:
+    if not isinstance(nn.Linear, module):
+        raise RuntimeError(
+            f"{module} is not a torch.nn.Linear module but _parallelize_linear was called!"
```

Comment:
we shouldn't disclose the function information in error message as we don't expose it to the user.  it should just be expecting torch.nn.Linear but found ...